Replicator/tests/ReplicatorWalrusTest.cc

```diff
@@ -675,6 +675,179 @@ TEST_CASE_METHOD(ReplicatorWalrusTest, "Pull iTunes deltas from SG", "[.SyncServ
           timeWithDelta, timeWithoutDelta, timeWithoutDelta/timeWithDelta);
 }
 
+
+TEST_CASE_METHOD(ReplicatorWalrusTest, "Pull deltas with filter from SG", "[.SyncServerWalrus][Delta]") {
+    static constexpr int kNumDocs = 10, kNumProps = 100;
+    flushScratchDatabase();
+    _logRemoteRequests = false;
+
+    // -------- Populating local db --------
+    auto populateDB = [&]() {
+        TransactionHelper t(db);
+        std::srand(123456); // start random() sequence at a known place
+        for (int docNo = 0; docNo < kNumDocs; ++docNo) {
+            char docID[kDocBufSize];
+            snprintf(docID, kDocBufSize, "doc-%03d", docNo);
+            Encoder enc(c4db_createFleeceEncoder(db));
+            enc.beginDict();
+            for (int p = 0; p < kNumProps; ++p) {
+                enc.writeKey(format("field%03d", p));
+                enc.writeInt(std::rand());
+            }
+            enc.endDict();
+            alloc_slice body = enc.finish();
+            string revID = createNewRev(db, slice(docID), body);
+        }
+    };
+    populateDB();
+
+    // -------- Pushing to SG --------"
+    replicate(kC4OneShot, kC4Disabled);
+
+    // -------- Updating docs on SG --------
+    {
+        JSONEncoder enc;
+        enc.beginDict();
+        enc.writeKey("docs"_sl);
+        enc.beginArray();
+        for (int docNo = 0; docNo < kNumDocs; ++docNo) {
+            char docID[kDocBufSize];
+            snprintf(docID, kDocBufSize, "doc-%03d", docNo);
+            C4Error error;
+            c4::ref<C4Document> doc = c4doc_get(db, slice(docID), false, ERROR_INFO(error));
+            REQUIRE(doc);
+            Dict props = c4doc_getProperties(doc);
+
+            enc.beginDict();
+            enc.writeKey("_id"_sl);
+            enc.writeString(docID);
+            enc.writeKey("_rev"_sl);
+            enc.writeString(doc->revID);
+            for (Dict::iterator i(props); i; ++i) {
+                enc.writeKey(i.keyString());
+                auto value = i.value().asInt();
+                if (RandomNumber() % 2 == 0)
+                    value = RandomNumber();
+                enc.writeInt(value);
+            }
+            enc.endDict();
+        }
+        enc.endArray();
+        enc.endDict();
+        _sg.sendRemoteRequest("POST", "_bulk_docs", enc.finish(), false, HTTPStatus::Created);
+    }
+
+    // -------- Repopulating local db --------
+    deleteAndRecreateDB();
+    populateDB();
+
+    // Setup pull filter:
+    _pullFilter = [](C4String collectionName, C4String docID, C4String revID,
+                     C4RevisionFlags flags, FLDict flbody, void *context) {
+        return true;
+    };
+
+    // -------- Pulling changes from SG --------
+    // CBL-4418: this test is interesting only if the puller finds an error of
+    // "Invalid delta". This error is not common. In the attachments of the ticket,
+    // a patch file, "cbl-4418-inject-error-pull.patch," can be found.
```

Comment:
Can it?  Wasn't this removed?  If so, is this test providing anything?

Reply:
I have't put them there yet. I thought to put them there after the PR is approved. Just put them there. Can always remove them if it's not a good idea.
Yes, this is a dilemma. I cannot find a way to generate the error in the test itself. With patch files, we can at least reproduce the current state sometime in the future.

Reply:
I don't much like adding tests that do nothing.  If they rely on the patch, then they are not useful in their current state and should be added along with the patch if and when it is added.

Reply:
So, you are suggesting to put the test cases and the code that inject the errors together into a patch and save it in the CBL ticket?

Reply:
I'm not sure that is a good long term solution, since what happens when the patch becomes unable to be applied?  I was thinking that if you had a patch you could add the test cases later and commit the patch file along side of it.  That also has the same problem though.  Could you create a special compile mode for this?  That way it won't suffer from patch rot and you can conditionally compile in the tests using the same define.

Reply:
Store the (combined) patch file in the Ticket has an advantage that we can alway replay the test, since the ticket also has the commit hash. To replay in the future, simply rollback the commit and apply the patch.
What do you think?